core/domain/prod_validation_jobs_one_off.py

```diff
@@ -0,0 +1,133 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""One-off jobs for validating prod models."""
+import logging
+
+from core import jobs
+from core.platform import models
+
+
+(
+    user_models, exp_models, collection_models,
+    feedback_models) = models.Registry.import_models([
+        models.NAMES.user, models.NAMES.exploration,
+        models.NAMES.collection, models.NAMES.feedback])
+
+INVALID_USER_SUBSCRIPTION_MODEL = 'invalid_user_subscription_model'
+VALID_USER_SUBSCRIPTION_MODEL = 'valid_user_subscription_model'
+
+
+
+def has_model_for_id(model_class, model_id):
+    """Checks there's an existing nondeleted model for given model_id.
+
+    Args:
+        model_class: BaseModel. Class of the model.
+        model_id: str. Key of the model.
+
+    Returns:
+        bool. True if there exists a model with given model_id.
+        """
+    model = model_class.get_by_id(model_id)
+    return model is not None and model.deleted is False
+
+
+class ValidateUserSubscriptionsModel(object):
+    """Validate UserSubscriptionsModels."""
+
+    errors = []
+    checks_passed = 0
+
+    @classmethod
+    def validate(cls, item):
+        """Validates a UserSubscriptionsModel entity.
+
+        Args:
+            item: UserSubscriptionsModel. Entity to validate.
+
+        Returns:
+            string. Validation status.
+        """
+        external_ids_checks = {
+            'activity_ids': exp_models.ExplorationModel,
+            'collection_ids': collection_models.CollectionModel,
+            'general_feedback_thread_ids':
+            feedback_models.GeneralFeedbackThreadModel,
+            'creator_ids': user_models.UserSettingsModel,
+        }
+        if not has_model_for_id(user_models.UserSettingsModel, item.id):
+            cls.errors.append(
+                'id %s is not valid %s' % (
+                    item.id,
+                    user_models.UserSettingsModel.__name__))
+        cls.check_external_ids(item, external_ids_checks=external_ids_checks)
+
+    @classmethod
+    def check_external_ids(cls, item, **kwargs):
+        """Check whether the external id properties on the model correspond
+        to valid instances.
+        """
+        external_ids_checks = kwargs.get('external_ids_checks', {})
+        for prop_name, model_class in external_ids_checks.iteritems():
+            model_ids = getattr(item, prop_name)
+            if not model_ids:
+                if model_ids is None:
+                    logging.warning(
+                        '%s is not a valid property on %s' %
+                        (prop_name, item.__class__.__name__))
+                continue
+
+            if isinstance(model_ids, str):
```

Comment:
This seems wrong, why would it be a string?

Also see comment above about using core.platform.datastore.

Reply:
Seems that datastore.fetch_multiple_entities_by_ids_and_models is used to fetch multiple models at once, but this check is for whether a model with given id exists or not. I guess fetch_multiple_entities_by_ids_and_models should be used only when we know the ids exist and want to do additional checks on the models.

Reply:
See above comment -- it might be returning None in locations which correspond to nonexistent models, and if that's the case, it's probably fine.